### PR TITLE
refine script to follow dcpy

### PR DESCRIPTION
**Pull request by SamYuan1990**
Changes for this PR, ref don't copy paste ourself
- make kind and microshift script just have cluster provide specific code.
- make common code in another files.
- some clean up.
- parameter able in test case.

### CODE REVIEW

**Review by vprashar2929 — CHANGES_REQUESTED** [review]

**Review by rootfs — APPROVED** [review]

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -12,14 +12,17 @@ on:
...
     runs-on: ubuntu-latest
     strategy:
       matrix:
         # I suppose we'd better update version pre kepler release.
         # keep from current default version and latest version.
+        platform: [kind, microshift]
         kubectl_version: [v1.25.4, v1.27.2]
+        # So far we just have a single version for microshift
+        # in the future, we should considering have a local test for each cluster providers
         kind_version: [0.17.0, 0.19.0]
```
Comment (marceloamaral) [inline .github/workflows/test.yml:26]: I still think we don't need to test different version that does not have major changes...

**Review by marceloamaral — APPROVED** [review]
/lgtm

File: .github/workflows/test.yml (hunk 1 above)
Comment (SamYuan1990) [reply .github/workflows/test.yml:26]: I suppose the depends on cluster provider.
for example micro shift don't have any other label?https://quay.io/repository/microshift/microshift-aio?tab=tags&tag=latest
or say the label logic looks wired.

File: kind/kind.sh [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ESTIMATOR_REPO=${ESTIMATOR_REPO:-quay.io/sustainable_computing_io}
+MODEL_SERVER_REPO=${MODEL_SERVER_REPO:-quay.io/sustainable_computing_io}
+
+CONFIG_OUT_DIR=${CONFIG_OUT_DIR:-"_output/generated-manifest"}
+KIND_DIR=${KIND_DIR:-"kind"}
+rm -rf ${CONFIG_OUT_DIR}
+mkdir -p ${CONFIG_OUT_DIR}
+
+# check CPU arch
+PLATFORM=$(uname -m)
```
[the remark this answers is not in the document]
Comment (SamYuan1990) [reply kind/kind.sh:44]: I am not sure where do we use PLATFORM?
in the currently logic we have PLATFORM points to either kind or micro shift or uanme -m ....
I suppose it's some mis take with current logic.
I have an idea in my mind that change PLATFORM to k8s_provider.

File: .github/workflows/test.yml (hunk 1 above)
Comment (marceloamaral) [reply .github/workflows/test.yml:26]: We could just have:
```
 kubectl_version: [v1.27.2]
        # So far we just have a single version for microshift
        # in the future, we should considering have a local test for each cluster providers
        kind_version: [0.19.0]
```

File: kind/kind.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (vprashar2929) [reply kind/kind.sh:44]: Yes, so what happens currently is that when we do `./main.sh up` and `export PLATFORM=microshift` then the `common.sh` uses the default case as the PLATFORM value gets changed by [this](https://github.com/sustainable-computing-io/local-dev-cluster/blob/24b060e9cd348287baf6675d847ac760f30c6802/kind/kind.sh#L44). and it compares on the basis of the arch, not actual PLATFORM env variable value. So it always set up the kind cluster.

[the remark this answers is not in the document]
Comment (vprashar2929) [reply kind/kind.sh:44]: I think changing the variable name to something like CPUArch will do the work.

**Review by marceloamaral — CHANGES_REQUESTED** [review]
units-test in microshift has failed.

The script should fix the following error:

```
++ kind get kubeconfig --name=kind
ERROR: could not locate any control plane nodes for cluster named 'kind'. Use the --name option to select a different cluster
fail to get kubeconfig by provider
+ '[' 0 == 0 ']'
+ echo 'fail to get kubeconfig by provider'
+ exit 1
Error: Process completed with exit code 1.
```

**Comment by SamYuan1990** [comment]
I will change this into WIP, and keep your guys noticed once ready.
I have some personal issue for now.

**Comment by SamYuan1990** [comment]
Updated, please help review and merge.

**Comment by vprashar2929** [comment]
We can also make kubectl version parameterized [here](https://github.com/sustainable-computing-io/local-dev-cluster/blob/f9be4b59e650667374961f4baf62726238beec1e/.github/workflows/test.yml#L54C12-L54C12)

**Comment by vprashar2929** [comment]
@SamYuan1990 I did some testing around this and I came to realize that in my [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/21) there is a bug that I introduced😞. If you try to deploy a kind cluster along with the registry and make kind cluster pull an image from the local registry then it fails to pull the image. I tested it now and IMO:

1. We have to use the `_registry_name="kind-registry"`, `REGISTRY_NAME=${REGISTRY_NAME:-kind-registry}` and `DEFAULT_NETWORK="kind"` in case we want to deploy kind cluster along with registry as without these kind is unable to connect with the registry.
2. This means that we have to make these variables separate for kind as well as micro shift when used in `common.sh`
3. We also have to move the `_run_registry` to the respective scripts.

**Comment by SamYuan1990** [comment]
> @SamYuan1990 I did some testing around this and I came to realize that in my [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/21) there is a bug that I introduced😞. If you try to deploy a kind cluster along with the registry and make kind cluster pull an image from the local registry then it fails to pull the image. I tested it now and IMO:
> 
> 1. We have to use the `_registry_name="kind-registry"`, `REGISTRY_NAME=${REGISTRY_NAME:-kind-registry}` and `DEFAULT_NETWORK="kind"` in case we want to deploy kind cluster along with registry as without these kind is unable to connect with the registry.
> 2. This means that we have to make these variables separate for kind as well as micro shift when used in `common.sh`
> 3. We also have to move the `_run_registry` to the respective scripts.

wait, in this pr, the kind is loaded image from local?
hence it's fixed in this PR?
https://github.com/sustainable-computing-io/local-dev-cluster/actions/runs/5365101787/jobs/9733716865?pr=20#step:4:737

**Comment by SamYuan1990** [comment]
> @SamYuan1990 I did some testing around this and I came to realize that in my [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/21) there is a bug that I introduced😞. If you try to deploy a kind cluster along with the registry and make kind cluster pull an image from the local registry then it fails to pull the image. I tested it now and IMO:
> 
> 1. We have to use the `_registry_name="kind-registry"`, `REGISTRY_NAME=${REGISTRY_NAME:-kind-registry}` and `DEFAULT_NETWORK="kind"` in case we want to deploy kind cluster along with registry as without these kind is unable to connect with the registry.
> 2. This means that we have to make these variables separate for kind as well as micro shift when used in `common.sh`
> 3. We also have to move the `_run_registry` to the respective scripts.

would you like to update with a test case, in case we can avoid this issue in the future ?

**Comment by vprashar2929** [comment]
> > @SamYuan1990 I did some testing around this and I came to realize that in my [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/21) there is a bug that I introduced😞. If you try to deploy a kind cluster along with the registry and make kind cluster pull an image from the local registry then it fails to pull the image. I tested it now and IMO:
> > 
> > 1. We have to use the `_registry_name="kind-registry"`, `REGISTRY_NAME=${REGISTRY_NAME:-kind-registry}` and `DEFAULT_NETWORK="kind"` in case we want to deploy kind cluster along with registry as without these kind is unable to connect with the registry.
> > 2. This means that we have to make these variables separate for kind as well as micro shift when used in `common.sh`
> > 3. We also have to move the `_run_registry` to the respective scripts.
> 
> wait, in this pr, the kind is loaded image from local? hence it's fixed in this PR? https://github.com/sustainable-computing-io/local-dev-cluster/actions/runs/5365101787/jobs/9733716865?pr=20#step:4:737

AFAIK `kind load` will load Docker images into cluster nodes. It doesn't use the local registry to push/pull images at the time of deployment. That's why CI works fine. How I tested was to push Kepler image to the local registry and then try to deploy Kepler using the image from the local registry. At that moment kind was not able to pull that image from the local registry. So I think the issue is still there in this PR.

**Comment by vprashar2929** [comment]
> > @SamYuan1990 I did some testing around this and I came to realize that in my [PR](https://github.com/sustainable-computing-io/local-dev-cluster/pull/21) there is a bug that I introduced😞. If you try to deploy a kind cluster along with the registry and make kind cluster pull an image from the local registry then it fails to pull the image. I tested it now and IMO:
> > 
> > 1. We have to use the `_registry_name="kind-registry"`, `REGISTRY_NAME=${REGISTRY_NAME:-kind-registry}` and `DEFAULT_NETWORK="kind"` in case we want to deploy kind cluster along with registry as without these kind is unable to connect with the registry.
> > 2. This means that we have to make these variables separate for kind as well as micro shift when used in `common.sh`
> > 3. We also have to move the `_run_registry` to the respective scripts.
> 
> would you like to update with a test case, in case we can avoid this issue in the future ?

Definitely what we can do is instead of using `kind load` we can add a step to push the images of grafana/prometheus-operator to the local registry and try to pull images from that at the time of deployment. This will ensure validation of local-registry wrt to kind/Microsoft cluster. Wdyt?

**Comment by SamYuan1990** [comment]
Personally, if design a case need more time to make it happen. @vprashar2929 , @marceloamaral and @rootfs I would like to open an issue to add a test case for load image test and merge current PR.

**Review by marceloamaral — APPROVED** [review]
/lgtm

**Pull request merged**